### ISSUE TITLE
[moved from data-services]  iso codes don't georeference consistently

### ISSUE THREAD

**Issue by iriberri** [CartoDB/cartodb#7386]
From @ztephm in https://github.com/CartoDB/data-services/issues/214

> Hi @iriberri can you help with this or please re-assign? Thanks!
>
>S/B 8621784
>
>## Context
>Dataset contains column of ISO3 codes. When importing and syncing same codes, sometimes they >georeference, sometimes they fail.
>
>## Process
>1. Create hourly sync-table with type & content-guessing box checked from [this GDrive file](https://drive.google.com/file/d/0Bw0ywRFon5y5a3luZ3pvcjJTcVU/view?usp=sharing)
>2. Create hourly sync-table with type & content-guessing box checked from [this GDrive file](https://drive.google.com/file/d/0Bw0ywRFon5y5M0hrQk5yRzMyWnc/view?usp=sharing) too 

>## Expected result
>They should both be automatically georeferenced the same way, since they contain the same ISO codes. 

>## Current result
>Here's [first one:](https://team.cartodb.com/u/stephaniemongon/tables/iso_only_1to10)

>![success_1to10](https://cloud.githubusercontent.com/assets/1779444/15193576/893cf662-178c-11e6-891f-8eaba0ef59fa.png)

>Here's [second one:](https://team.cartodb.com/u/stephaniemongon/tables/iso_only_1to11)
>
>![fail_1to11](https://cloud.githubusercontent.com/assets/1779444/15193601/a86e907c-178c-11e6-86d8-c49e6255a796.png)

>Anything we can do to make sure they always georeference properly?

**Comment by iriberri** [comment]
First of all, the behaviour above is consistent with the current configuration of the system.

Geocoding guessing strongly depends on the entropy of the data that is in the table.

For each text column the system calculates the entropy of its data:

````
      # See http://en.wikipedia.org/wiki/Entropy_(information_theory)
      # See http://www.shannonentropy.netmark.pl/
      #
      # Returns 0.0 if all elements in the column are repeated
      # Returns 1.0 if all elements in the column are different
      def metric_entropy(column, normalizer=nil)
        shannon_entropy(column, normalizer) / Math.log(sample.count)
      end

      def shannon_entropy(column, normalizer)
        sum = 0.0
        frequencies(column, normalizer).each { |freq| sum += (freq * Math.log(freq)) }
        return sum.abs
      end

      # Returns an array with the relative frequencies of the elements of that column
      def frequencies(column, normalizer)
        frequency_table = {}
        column_name_sym = column[:column_name].to_sym
        if normalizer
          sample.each do |row|
            elem = normalizer.call(row[column_name_sym])
            frequency_table[elem] += 1 rescue frequency_table[elem] = 1
          end
        else
          sample.each do |row|
            elem = row[column_name_sym]
            frequency_table[elem] += 1 rescue frequency_table[elem] = 1
          end
        end
        length = sample.count.to_f
        frequency_table.map { |key, value| value / length }
      end
````
From: https://github.com/CartoDB/cartodb/blob/ebfb0e62bfdb6bd941408b4d2f0826b3195318fb/services/importer/lib/importer/content_guesser.rb#L105-L137

The entropy value computed can go from 0 up to 1, and the value depends on how much the data is repeated in the column. If this value turns out to be smaller than a threshold, then the system decides that the data does not require to be guessed and the dataset doesn't get georeferenced on import. This threshold value, or minimum entropy is set to 0.9.

For a table of 10 rows, having 5 repeated rows is going to make the entropy value small. In a table of 1000 rows, 5 repeated rows wouldn't have that effect in the entropy computation.

So what is happening here is that the table is pretty small, and it has several repeated values. The difference from the table with 9 rows to the table with 10 rows is a new row which indeed contains repeated data (BFA already appeared in other row, and BDI is also repeated).

The idea of computing the entropy is that it doesn't make sense to just geocode all the data. Imagine a table with 10 rows for countries, in which "France" appears 5 times. Would it make sense to have 5 overlapping polygons there? It wouldn't, right? Then the entropy computation would avoid the dataset to be geocoded.

For these particular datasets, here are the entropy values for the iso3 column:

9 rows: 0.9298966940476159
10 rows: 0.8795880017344073
minimum_entropy: 0.9

For 10 rows, 0.87 < 0.9, so it doesn't get guessed.

So the conclusion is: the dataset is not being geocoded because there's a lot of repeated data for such a small dataset. Is this just a test dataset? With more data, as the sample increases, the guessing wouldn't be so sensible to row repetitions.

Hope this helps!

**Comment by remofambri** [comment]
So is the solution to add more countries to the list to make at least 90%? 
Assuming that we actually need duplicates...

**Comment by iriberri** [comment]
If you are using the SaaS from cartodb.com, the minimum entropy is fixed to 0.9, so yes, the guessing would only be triggered if your data has an entropy >= 0.9.

If more data is added (and this new data doesn't contain a low of new duplicated values) the entropy value will increase and it will make your use case work. Removing the duplicates, by trying to join their data in your original dataset (if possible) will also improve this value.

Unfortunately, this use case doesn't seem the typical example for which the guessing was designed, which is always taking into account how "mappable" would the output geometries be. 

It's not nice to add random data to the dataset which you don't need (although, as a super hack, adding rows which do not correspond to a geocodificable country would also work, as it increases the entropy but this data would not make it to the final map because it won't be geocoded). Perhaps you can also consider the option of geocoding your data by using a join between your dataset and a country borders dataset to obtain the geometries and joining the ISO codes of both.
